portal/src/it/scala/cromwell/pipeline/controller/AuthControllerItTest.scala

```diff
@@ -0,0 +1,100 @@
+package cromwell.pipeline.controller
+
+import java.util.UUID
+
+import akka.http.scaladsl.model.ContentTypes.`application/json`
+import akka.http.scaladsl.model.{HttpEntity, StatusCodes}
+import akka.http.scaladsl.testkit.ScalatestRouteTest
+import com.dimafeng.testcontainers.ForAllTestContainer
+import cromwell.pipeline.BaseItTest
+import cromwell.pipeline.controller.AuthController._
+import cromwell.pipeline.datastorage.dto.auth.{SignInRequest, SignUpRequest}
+import cromwell.pipeline.datastorage.dto.{User, UserId}
+import cromwell.pipeline.utils.StringUtils
+import org.scalatest.compatible.Assertion
+import play.api.libs.json.Json
+
+
+class AuthControllerItTest extends BaseItTest with ScalatestRouteTest with ForAllTestContainer {
```

Comment:
I'm a scala newbie, but from the OOP point it's not a good practice to use inheritance in tests. It'd be great to replace it with composition if possible. Here is the first link given by google (for Java) https://javarevisited.blogspot.com/2013/06/why-favor-composition-over-inheritance-java-oops-design.html

Reply:
@kynyan, thank you for your proposal,

I've investigated the article on which you refer but noticed only one thesis about tests (where the mocking process was described). Here we have integration testing and we do not need to mock something.

With respect to our unit tests, we can easily mock other components and I don't see any reason to change something in current logic.

Could you please describe your suggestion in more detail what exact changes you would like to add?

Reply:
Ok, I'm not sure if I can find anything specifically about tests, may be [this one](https://www.petrikainulainen.net/programming/unit-testing/3-reasons-why-we-should-not-use-inheritance-in-our-tests/) is closer. And it's hard to give a concrete example for this project, but it all comes down to the limitations introduced by inheritance. Once you extend your test from something and then you need to change some part of what you inherit for another test, you are going to make duplicated code, and there is also a risk of breaking child's functionality. 

I've heard "the rule of thumb": use inheritance to model **is-a** relationships, and composition to model **has-a** relationships. `AuthControllerItTest` just uses (**has**) access to the DB which is encapsulated in `BaseTest`. In Java I would create annotation `@RepositoryTest` which would enable DB config for the test, but I don't know if it can be done with Scala.

Reply:
I agree with @kynyan. We should try to use composition instead of inheritance. 
At least, maybe I will need `beforeEach` to have other behaviour than defined right now in the `BaseItTest`. So I will need to override something that you've already overrode. So what is the point.

Reply:
@kynyan, yes, the second article is closer and I've got what you're talking about. Thank you!

You can check the updated version when you will have a few minutes.

Reply:
Looks good to me, thanx!